apm-lambda-extension/extension/apm_server_test.go

```diff
@@ -94,13 +90,204 @@ func TestPostToApmServerDataNotCompressed(t *testing.T) {
+	go func() {
+		gw.Write([]byte(""))
+		gw.Close()
+		pw.Close()
+	}()
+
+	// Create AgentData struct with compressed data
+	data, _ := ioutil.ReadAll(pr)
+	agentData := AgentData{Data: data, ContentEncoding: "gzip"}
+
+	// Create apm server and handler
+	apmServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		bytes, _ := ioutil.ReadAll(r.Body)
+		assert.Equal(t, string(data), string(bytes))
+		assert.Equal(t, "gzip", r.Header.Get("Content-Encoding"))
+		w.Write([]byte(`{"foo": "bar"}`))
+	}))
+	apmServer.Close()
+
+	config := extensionConfig{
+		apmServerUrl: apmServer.URL + "/",
+	}
+
+	PostToApmServer(apmServer.Client(), agentData, &config, context.Background())
+	assert.Equal(t, apmServerTransportStatus, TransportFailing)
+	assert.Equal(t, apmServerReconnectionCount, 1)
+}
+
+func TestAPMServerRecovery(t *testing.T) {
+	SetApmServerTransportStatus(TransportHealthy, 0)
+	enterBackoff(context.Background())
+	for {
+		if IsTransportStatusHealthy() {
+			break
+		}
+	}
+	assert.Equal(t, apmServerTransportStatus, TransportPending)
+
+	// Compress the data
+	pr, pw := io.Pipe()
+	gw, _ := gzip.NewWriterLevel(pw, gzip.BestSpeed)
+	go func() {
+		gw.Write([]byte(""))
+		gw.Close()
+		pw.Close()
+	}()
+
+	// Create AgentData struct with compressed data
+	data, _ := ioutil.ReadAll(pr)
+	agentData := AgentData{Data: data, ContentEncoding: "gzip"}
+
+	// Create apm server and handler
+	apmServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		bytes, _ := ioutil.ReadAll(r.Body)
+		assert.Equal(t, string(data), string(bytes))
+		assert.Equal(t, "gzip", r.Header.Get("Content-Encoding"))
+		w.Write([]byte(`{"foo": "bar"}`))
+	}))
+	defer apmServer.Close()
+
+	config := extensionConfig{
+		apmServerUrl: apmServer.URL + "/",
+	}
+
+	PostToApmServer(apmServer.Client(), agentData, &config, context.Background())
+
+	assert.Equal(t, apmServerTransportStatus, TransportHealthy)
+	assert.Equal(t, apmServerReconnectionCount, 0)
+}
+
+func TestContinuedAPMServerFailure(t *testing.T) {
+	SetApmServerTransportStatus(TransportHealthy, 0)
+	enterBackoff(context.Background())
+
+	for {
+		if IsTransportStatusHealthy() {
+			break
+		}
+	}
+
+	assert.Equal(t, apmServerTransportStatus, TransportPending)
+
+	// Compress the data
+	pr, pw := io.Pipe()
+	gw, _ := gzip.NewWriterLevel(pw, gzip.BestSpeed)
+	go func() {
+		gw.Write([]byte(""))
+		gw.Close()
+		pw.Close()
+	}()
+
+	// Create AgentData struct with compressed data
+	data, _ := ioutil.ReadAll(pr)
+	agentData := AgentData{Data: data, ContentEncoding: "gzip"}
+
+	// Create apm server and handler
+	apmServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		bytes, _ := ioutil.ReadAll(r.Body)
+		assert.Equal(t, string(data), string(bytes))
+		assert.Equal(t, "gzip", r.Header.Get("Content-Encoding"))
+		w.Write([]byte(`{"foo": "bar"}`))
+	}))
+	apmServer.Close()
```

Comment:
and call out that you're closing the server here

Reply:
Solved in 240fa40